File: src/bin/tui/telnet.rs
//! The client half of `src/core/network/telnet/`.
//!
//! `TelnetParser` is a pure byte state machine — it does not care which end of
//! the connection it is on — so the protocol decoding here is the driver's own
//! code, and only the negotiation *answers* are new.
//!
//! The server offers `WILL SGA`, `DO SGA`, `WILL GMCP`, `WILL MCCP2`, `DO TTYPE`
//! and `DO NAWS` on connect (`connection.rs::send_initial_negotiations`), then
//! `WILL ECHO` / `WONT ECHO` around the password prompt.

use std::collections::HashSet;

use oxigeon::core::network::telnet::codec::TelnetCodec;
use oxigeon::core::network::telnet::constants::*;
use oxigeon::core::network::telnet::parser::{TelnetEvent, TelnetParser};
use tokio::io::{AsyncReadExt, AsyncWriteExt};
use tokio::net::TcpStream;
use tokio::sync::mpsc::{UnboundedReceiver, UnboundedSender};

use crate::ansi::AnsiDecoder;
use crate::app::{Action, AppEvent};

/// What we tell the server we understand. `gmcp_d.wants()` gates every outbound
/// package on this list — a module covers its packages, so `Char` buys
/// `Char.Vitals`, `Char.Status` and `Char.Effects`.
const SUPPORTS: &str = r#"["Char 1","Room 1","Core 1"]"#;

const TERMINAL_TYPE: &str = "OXIGEON-TUI";

/// TTYPE subnegotiation commands (RFC 1091).
const TTYPE_IS: u8 = 0;
const TTYPE_SEND: u8 = 1;

pub async fn run(
    addr: String,
    events: UnboundedSender<AppEvent>,
    mut actions: UnboundedReceiver<Action>,
) {
    let stream = match TcpStream::connect(&addr).await {
        Ok(s) => s,
        Err(e) => {
            let _ = events.send(AppEvent::TelnetDown(format!("{}: {}", addr, e)));
            return;
        }
    };
    let _ = stream.set_nodelay(true);
    let _ = events.send(AppEvent::TelnetUp);

    let (mut reader, mut writer) = stream.into_split();
    let mut parser = TelnetParser::new();
    let mut ansi = AnsiDecoder::new();
    let mut answered: HashSet<(u8, u8)> = HashSet::new();
    let mut size = (100u16, 30u16);
    let mut buf = [0u8; 8192];

    loop {
        tokio::select! {
            read = reader.read(&mut buf) => {
                let n = match read {
                    Ok(0) => {
                        let _ = events.send(AppEvent::TelnetDown("server closed".into()));
                        return;
                    }
                    Ok(n) => n,
                    Err(e) => {
                        let _ = events.send(AppEvent::TelnetDown(e.to_string()));
                        return;
                    }
                };
                let mut out: Vec<u8> = Vec::new();
                for event in parser.feed_bytes(&buf[..n]) {
                    handle(event, &events, &mut ansi, &mut answered, &mut out, size);
                }
                if !out.is_empty() && writer.write_all(&out).await.is_err() {
                    let _ = events.send(AppEvent::TelnetDown("write failed".into()));
                    return;
                }
            }
            action = actions.recv() => {
                let bytes = match action {
                    None => return,
                    Some(Action::Send(line)) => {
                        // encode_text handles LF → CRLF and escapes 0xFF as IAC IAC,
                        // so a player typing a high byte cannot inject a command.
                        TelnetCodec::encode_text(&format!("{}\n", line))
                    }
                    Some(Action::Naws(w, h)) => {
                        size = (w, h);
                        naws(w, h)
                    }
                    Some(Action::Dap(..)) => continue,
                };
                if writer.write_all(&bytes).await.is_err() {
                    let _ = events.send(AppEvent::TelnetDown("write failed".into()));
                    return;
                }
            }
        }
    }
}

fn handle(
    event: TelnetEvent,
    events: &UnboundedSender<AppEvent>,
    ansi: &mut AnsiDecoder,
    answered: &mut HashSet<(u8, u8)>,
    out: &mut Vec<u8>,
    size: (u16, u16),
) {
    match event {
        TelnetEvent::Data(bytes) => {
            for line in ansi.feed(&bytes) {
                let _ = events.send(AppEvent::GameLine(line));
            }
            // Whatever is left unterminated is the prompt. The driver sends it
            // with no newline, which is exactly how we tell them apart.
            let _ = events.send(match ansi.partial() {
                Some(line) => AppEvent::GamePrompt(line),
                None => AppEvent::GamePrompt(ratatui::text::Line::default()),
            });
        }

        TelnetEvent::Negotiate { verb, option } => {
            // ECHO is not an ordinary option: the server toggles it repeatedly
            // around every password prompt, so it must never be de-duplicated.
            if option == OPT_ECHO {
                match verb {
                    WILL => {
                        out.extend(TelnetCodec::encode_negotiate(DO, OPT_ECHO));
                        let _ = events.send(AppEvent::Echo(true));
                    }
                    WONT => {
                        out.extend(TelnetCodec::encode_negotiate(DONT, OPT_ECHO));
                        let _ = events.send(AppEvent::Echo(false));
                    }
                    _ => {}
                }
                return;
            }

            // Everything else is answered once. Replying to a repeat is how a
            // naive client and a Q-method server talk each other into a loop.
            if !answered.insert((verb, option)) {
                return;
            }

            match (verb, option) {
                (WILL, OPT_GMCP) => {
                    out.extend(TelnetCodec::encode_negotiate(DO, OPT_GMCP));
                    // Identify, then ask for exactly the packages we render.
                    out.extend(TelnetCodec::encode_gmcp(
                        "Core.Hello",
                        Some(&format!(
                            r#"{{"client":"oxigeon-tui","version":"{}"}}"#,
                            env!("CARGO_PKG_VERSION")
                        )),
                    ));
                    out.extend(TelnetCodec::encode_gmcp(
                        "Core.Supports.Set",
                        Some(SUPPORTS),
                    ));
                }

                // MCCP2 is negotiated by the driver but never performed —
                // `flate2` appears nowhere in `src/` and `mccp2_active` is never
                // set. Accepting it would agree to a compression that never
                // starts, and every byte after would be garbage.
                (WILL, OPT_MCCP2) | (WILL, OPT_MCCP3) => {
                    out.extend(TelnetCodec::encode_negotiate(DONT, option));
                }

                (DO, OPT_NAWS) => {
                    out.extend(TelnetCodec::encode_negotiate(WILL, OPT_NAWS));
                    out.extend(naws(size.0, size.1));
                }
                (DO, OPT_TTYPE) => out.extend(TelnetCodec::encode_negotiate(WILL, OPT_TTYPE)),
                (WILL, OPT_SGA) => out.extend(TelnetCodec::encode_negotiate(DO, OPT_SGA)),
                (DO, OPT_SGA) => out.extend(TelnetCodec::encode_negotiate(WILL, OPT_SGA)),

                // Refuse anything we have not implemented, rather than leaving
                // it unanswered — an unanswered DO stalls some servers.
                (WILL, opt) => out.extend(TelnetCodec::encode_negotiate(DONT, opt)),
                (DO, opt) => out.extend(TelnetCodec::encode_negotiate(WONT, opt)),
                _ => {}
            }
        }

        TelnetEvent::Subnegotiation { option, data } => match option {
            OPT_GMCP => {
                if let Some((package, json)) = TelnetCodec::parse_gmcp(&data) {
                    let _ = events.send(AppEvent::Gmcp {
                        package,
                        json: json.unwrap_or_else(|| "null".into()),
                    });
                }
            }
            OPT_TTYPE if data.first() == Some(&TTYPE_SEND) => {
                let mut payload = vec![TTYPE_IS];
                payload.extend_from_slice(TERMINAL_TYPE.as_bytes());
                out.extend(TelnetCodec::encode_subnegotiation(OPT_TTYPE, &payload));
            }
            _ => {}
        },

        // NOP, AYT and friends. Nothing here needs a response the driver acts on.
        TelnetEvent::Command(_) => {}
    }
}

/// NAWS payload is two big-endian u16s: width then height.
fn naws(w: u16, h: u16) -> Vec<u8> {
    let payload = [(w >> 8) as u8, w as u8, (h >> 8) as u8, h as u8];
    TelnetCodec::encode_subnegotiation(OPT_NAWS, &payload)
}

#[cfg(test)]
mod tests {
    use super::*;
    use tokio::sync::mpsc;

    fn drive(input: &[u8]) -> (Vec<u8>, Vec<AppEvent>) {
        let (tx, mut rx) = mpsc::unbounded_channel();
        let mut parser = TelnetParser::new();
        let mut ansi = AnsiDecoder::new();
        let mut answered = HashSet::new();
        let mut out = Vec::new();
        for event in parser.feed_bytes(input) {
            handle(event, &tx, &mut ansi, &mut answered, &mut out, (80, 24));
        }
        drop(tx);
        let mut events = Vec::new();
        while let Ok(e) = rx.try_recv() {
            events.push(e);
        }
        (out, events)
    }

    #[test]
    fn gmcp_offer_is_accepted_and_answered_with_hello_and_supports() {
        let (out, _) = drive(&[IAC, WILL, OPT_GMCP]);
        assert!(out.starts_with(&[IAC, DO, OPT_GMCP]));
        let text = String::from_utf8_lossy(&out);
        assert!(text.contains("Core.Hello"));
        assert!(text.contains("Core.Supports.Set"));
        assert!(text.contains("Char 1"));
    }

    #[test]
    fn mccp2_is_refused_because_the_driver_never_actually_compresses() {
        let (out, _) = drive(&[IAC, WILL, OPT_MCCP2]);
        assert_eq!(out, vec![IAC, DONT, OPT_MCCP2]);
    }

    #[test]
    fn echo_toggles_masking_in_both_directions_and_is_never_deduplicated() {
        // The driver toggles ECHO around every password prompt, so the second
        // WILL must still mask. De-duplicating it would silently expose a
        // password on a re-login.
        let (out, events) = drive(&[
            IAC, WILL, OPT_ECHO, IAC, WONT, OPT_ECHO, IAC, WILL, OPT_ECHO,
        ]);
        assert_eq!(
            out,
            vec![IAC, DO, OPT_ECHO, IAC, DONT, OPT_ECHO, IAC, DO, OPT_ECHO]
        );
        let masking: Vec<bool> = events
            .iter()
            .filter_map(|e| match e {
                AppEvent::Echo(on) => Some(*on),
                _ => None,
            })
            .collect();
        assert_eq!(masking, vec![true, false, true]);
    }

    #[test]
    fn naws_is_offered_with_the_current_size() {
        let (out, _) = drive(&[IAC, DO, OPT_NAWS]);
        assert!(out.starts_with(&[IAC, WILL, OPT_NAWS]));
        // 80x24 big-endian, wrapped in a subnegotiation.
        assert!(out.ends_with(&[IAC, SB, OPT_NAWS, 0, 80, 0, 24, IAC, SE]));
    }

    #[test]
    fn terminal_type_is_reported_when_asked() {
        let (out, _) = drive(&[IAC, SB, OPT_TTYPE, TTYPE_SEND, IAC, SE]);
        assert_eq!(out[..3], [IAC, SB, OPT_TTYPE]);
        assert_eq!(out[3], TTYPE_IS);
        assert!(String::from_utf8_lossy(&out).contains(TERMINAL_TYPE));
    }

    #[test]
    fn an_unknown_option_is_refused_rather_than_ignored() {
        let (out, _) = drive(&[IAC, WILL, 99, IAC, DO, 98]);
        assert_eq!(out, vec![IAC, DONT, 99, IAC, WONT, 98]);
    }

    #[test]
    fn a_repeated_offer_is_answered_once() {
        let (out, _) = drive(&[IAC, WILL, OPT_SGA, IAC, WILL, OPT_SGA]);
        assert_eq!(out, vec![IAC, DO, OPT_SGA]);
    }

    #[test]
    fn game_text_and_gmcp_interleaved_in_one_read_both_arrive() {
        let mut input = b"You see a well.\r\n".to_vec();
        input.extend(TelnetCodec::encode_gmcp(
            "Char.Vitals",
            Some(r#"{"hp":42,"maxhp":50}"#),
        ));
        input.extend_from_slice(b"42h> ");
        let (_, events) = drive(&input);

        assert!(events.iter().any(|e| matches!(
            e, AppEvent::GameLine(l) if l.spans.iter().any(|s| s.content.contains("well"))
        )));
        assert!(events.iter().any(|e| matches!(
            e, AppEvent::Gmcp { package, json } if package == "Char.Vitals" && json.contains("42")
        )));
        assert!(events.iter().any(|e| matches!(
            e, AppEvent::GamePrompt(l) if l.spans.iter().any(|s| s.content.contains("42h>"))
        )));
    }
}

File: debug-client/bridge/telnet.js
// Telnet, the client half.
//
// A port of `src/core/network/telnet/parser.rs` (a pure byte state machine, so
// it does not care which end of the connection it is on) plus the negotiation
// *answers* from `src/bin/tui/telnet.rs`.
//
// The server offers `WILL SGA`, `DO SGA`, `WILL GMCP`, `WILL MCCP2`, `DO TTYPE`
// and `DO NAWS` on connect (`connection.rs::send_initial_negotiations`), then
// `WILL ECHO` / `WONT ECHO` around the password prompt.

export const IAC = 255
export const SE = 240
export const SB = 250
export const WILL = 251
export const WONT = 252
export const DO = 253
export const DONT = 254

export const OPT_ECHO = 1
export const OPT_SGA = 3
export const OPT_TTYPE = 24
export const OPT_NAWS = 31
export const OPT_MCCP2 = 86
export const OPT_MCCP3 = 87
export const OPT_GMCP = 201

const TTYPE_IS = 0
const TTYPE_SEND = 1

/// What we tell the server we understand. `gmcp_d.wants()` gates every outbound
/// package on this list — a module covers its packages, so `Char` buys
/// `Char.Vitals`, `Char.Status` and `Char.Effects`.
const SUPPORTS = '["Char 1","Room 1","Core 1"]'
const TERMINAL_TYPE = 'OXIGEON-WEB'
const CLIENT_VERSION = '0.1.0'

// ─── codec ───────────────────────────────────────────────────────────────────

export function encodeNegotiate(verb, option) {
  return Buffer.from([IAC, verb, option])
}

export function encodeSubnegotiation(option, payload) {
  return Buffer.concat([
    Buffer.from([IAC, SB, option]),
    Buffer.from(payload),
    Buffer.from([IAC, SE]),
  ])
}

export function encodeGmcp(pkg, json) {
  const payload =
    json === undefined || json === null
      ? Buffer.from(pkg, 'utf8')
      : Buffer.concat([Buffer.from(pkg + ' ', 'utf8'), Buffer.from(json, 'utf8')])
  return encodeSubnegotiation(OPT_GMCP, payload)
}

/// Split a GMCP payload on the first space, as `TelnetCodec::parse_gmcp` does.
export function parseGmcp(data) {
  const at = data.indexOf(0x20)
  if (at === -1) return { package: data.toString('utf8'), json: null }
  return {
    package: data.subarray(0, at).toString('utf8'),
    json: at + 1 < data.length ? data.subarray(at + 1).toString('utf8') : null,
  }
}

/// LF → CRLF, and 0xFF escaped as IAC IAC so a player typing a high byte
/// cannot inject a command.
export function encodeText(text) {
  const out = []
  for (const byte of Buffer.from(text, 'utf8')) {
    if (byte === IAC) out.push(IAC, IAC)
    else if (byte === 0x0a) out.push(0x0d, 0x0a)
    else out.push(byte)
  }
  return Buffer.from(out)
}

/// NAWS payload is two big-endian u16s: width then height.
export function naws(w, h) {
  return encodeSubnegotiation(OPT_NAWS, [(w >> 8) & 0xff, w & 0xff, (h >> 8) & 0xff, h & 0xff])
}

// ─── parser ──────────────────────────────────────────────────────────────────

const NORMAL = 0
const IN_IAC = 1
const IN_NEGOTIATION = 2
const IN_SUB = 3
const IN_SUB_IAC = 4

/// Byte-level telnet parser (RFC 854). Feed it chunks; it yields events.
/// `data` events carry a Buffer, because a UTF-8 character can straddle a TCP
/// read and decoding here would corrupt it.
export class TelnetParser {
  constructor() {
    this.state = NORMAL
    this.verb = 0
    this.option = 0
    this.data = []
    this.sub = []
    this.subOptionPending = false
  }

  #flushData(out) {
    if (this.data.length) {
      out.push({ kind: 'data', bytes: Buffer.from(this.data) })
      this.data = []
    }
  }

  feed(chunk) {
    const out = []
    for (const byte of chunk) {
      switch (this.state) {
        case NORMAL:
          if (byte === IAC) {
            this.#flushData(out)
            this.state = IN_IAC
          } else {
            this.data.push(byte)
          }
          break

        case IN_IAC:
          if (byte === IAC) {
            // Escaped IAC — an ordinary 0xFF data byte.
            this.data.push(255)
            this.state = NORMAL
          } else if (byte === WILL || byte === WONT || byte === DO || byte === DONT) {
            this.verb = byte
            this.state = IN_NEGOTIATION
          } else if (byte === SB) {
            this.sub = []
            this.subOptionPending = true
            this.state = IN_SUB
          } else if (byte === SE) {
            this.state = NORMAL // SE without SB — ignore
          } else {
            out.push({ kind: 'command', command: byte })
            this.state = NORMAL
          }
          break

        case IN_NEGOTIATION:
          out.push({ kind: 'negotiate', verb: this.verb, option: byte })
          this.state = NORMAL
          break

        case IN_SUB:
          if (this.subOptionPending) {
            this.option = byte
            this.subOptionPending = false
          } else if (byte === IAC) {
            this.state = IN_SUB_IAC
          } else {
            this.sub.push(byte)
          }
          break

        case IN_SUB_IAC:
          if (byte === SE) {
            out.push({ kind: 'subnegotiation', option: this.option, data: Buffer.from(this.sub) })
            this.sub = []
            this.state = NORMAL
          } else if (byte === IAC) {
            this.sub.push(255)
            this.state = IN_SUB
          } else {
            // A stray command inside a subnegotiation. Abandon it rather than
            // swallowing the rest of the stream looking for an SE.
            this.state = IN_SUB
          }
          break
      }
    }
    this.#flushData(out)
    return out
  }
}

// ─── negotiation ─────────────────────────────────────────────────────────────

/// Answer one telnet event.
///
/// Returns `{ reply, emit }` — bytes to write back, and what the UI should be
/// told. `answered` is a Set the caller keeps across the connection: everything
/// except ECHO is answered exactly once.
export function respond(event, answered, size) {
  const reply = []
  const emit = []

  switch (event.kind) {
    case 'data':
      emit.push({ t: 'game', bytes: event.bytes })
      break

    case 'negotiate': {
      const { verb, option } = event

      // ECHO is not an ordinary option: the server toggles it repeatedly around
      // every password prompt, so it must never be de-duplicated.
      if (option === OPT_ECHO) {
        if (verb === WILL) {
          reply.push(encodeNegotiate(DO, OPT_ECHO))
          emit.push({ t: 'echo', on: true })
        } else if (verb === WONT) {
          reply.push(encodeNegotiate(DONT, OPT_ECHO))
          emit.push({ t: 'echo', on: false })
        }
        break
      }

      // Everything else is answered once. Replying to a repeat is how a naive
      // client and a Q-method server talk each other into a loop.
      const key = verb * 256 + option
      if (answered.has(key)) break
      answered.add(key)

      if (verb === WILL && option === OPT_GMCP) {
        reply.push(encodeNegotiate(DO, OPT_GMCP))
        reply.push(
          encodeGmcp('Core.Hello', JSON.stringify({ client: 'oxigeon-web', version: CLIENT_VERSION }))
        )
        reply.push(encodeGmcp('Core.Supports.Set', SUPPORTS))
      } else if (verb === WILL && (option === OPT_MCCP2 || option === OPT_MCCP3)) {
        // MCCP2 is negotiated by the driver but never performed. `flate2` is a
        // declared dependency and is used nowhere in `src/`; `mccp2_active` is
        // declared on the connection and never set. Accepting would agree to a
        // compression that never starts, and every byte after would be garbage.
        reply.push(encodeNegotiate(DONT, option))
      } else if (verb === DO && option === OPT_NAWS) {
        reply.push(encodeNegotiate(WILL, OPT_NAWS))
        reply.push(naws(size.w, size.h))
      } else if (verb === DO && option === OPT_TTYPE) {
        reply.push(encodeNegotiate(WILL, OPT_TTYPE))
      } else if (verb === WILL && option === OPT_SGA) {
        reply.push(encodeNegotiate(DO, OPT_SGA))
      } else if (verb === DO && option === OPT_SGA) {
        reply.push(encodeNegotiate(WILL, OPT_SGA))
      } else if (verb === WILL) {
        // Refuse anything we have not implemented, rather than leaving it
        // unanswered — an unanswered DO stalls some servers.
        reply.push(encodeNegotiate(DONT, option))
      } else if (verb === DO) {
        reply.push(encodeNegotiate(WONT, option))
      }
      break
    }

    case 'subnegotiation':
      if (event.option === OPT_GMCP) {
        const { package: pkg, json } = parseGmcp(event.data)
        emit.push({ t: 'gmcp', package: pkg, json: json ?? 'null' })
      } else if (event.option === OPT_TTYPE && event.data[0] === TTYPE_SEND) {
        reply.push(
          encodeSubnegotiation(OPT_TTYPE, Buffer.concat([Buffer.from([TTYPE_IS]), Buffer.from(TERMINAL_TYPE)]))
        )
      }
      break

    // NOP, AYT and friends. Nothing here needs a response the driver acts on.
    case 'command':
      break
  }

  return { reply: reply.length ? Buffer.concat(reply) : null, emit }
}
